src/components/TodosProvider/TodosProvider.tsx

```diff
@@ -0,0 +1,209 @@
+import React, {
+  createContext,
+  useContext,
+  FC,
+  useState,
+  useEffect,
+  useCallback,
+} from 'react';
+import { Todo } from '../../types/Todo';
+import { FilterStatus } from '../../types/FilterStatus';
+import { getTodos } from '../../api/todos';
+import * as todoService from '../../api/todos';
+import { errorMessages } from '../ErrorNotification';
+import { TodosContextType } from '../../types/contextTypes';
+import {
+  initialFilterStatus,
+  initialTodos,
+} from '../../types/initialContextValues';
+import { getFilteredTodos } from '../../utils/getFilteredTodos';
+
+interface Props {
+  children: React.ReactNode;
+}
+
+const TodosContext = createContext<TodosContextType>({
+  todos: initialTodos,
+  setTodos: () => {},
+  filterStatus: initialFilterStatus,
+  setFilterStatus: () => {},
+  filteredTodos: [],
+  isLoading: false,
+  errorMessage: '',
+  setErrorMessage: () => {},
+  isLoadingData: false,
+  setIsLoading: () => {},
+  selectedTodoIds: [],
+  setSelectedTodoIds: () => {},
+  tempTodo: null,
+  setTempTodo: () => {},
+  createNewTodo: () => {},
+  newTodoTitle: '',
+  setNewTodoTitle: () => {},
+  deleteTodo: () => {},
+  clearCompletedTodos: () => {},
+  updateTodo: () => {},
+});
+
+export const useTodos = () => useContext(TodosContext);
+
+export const TodosProvider: FC<Props> = ({ children }) => {
+  const [todos, setTodos] = useState<Todo[]>(initialTodos);
+
+  const [selectedTodoIds, setSelectedTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+
+  const [filterStatus, setFilterStatus] =
+    useState<FilterStatus>(initialFilterStatus);
+
+  const [isLoadingData, setIsLoadingData] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
```

Comment:
> Instead of setting `isLoading` you can create a state with an array of loading todo ids. So you can add and remove ids of todos that currently loading and show loaders for them.
> 
> As if I click "Clear completed" I don't see loaders on todos that are currently loading
> 
> UPD. I see that you have "selectedTodoIds", so it looks that you have too much loading states

You still have 3 states for loaders 👀